core/pythonAction/pythonrunner.py

```diff
@@ -36,23 +36,33 @@ def initCodeFromString(self, message):
 
     def build(self, message):
         binary = message['binary'] if 'binary' in message else False
-
         if not binary:
             code = message['code']
             filename = 'action'
         elif os.path.isfile(self.source):
             with codecs.open(self.source, 'r', 'utf-8') as m:
                 code = m.read()
-            filename = '__main__.py'
+            filename = os.path.basename(self.source)
             sys.path.insert(0, os.path.dirname(self.source))
         else:
-            sys.stderr.write('Zip file does not include "__main__.py".\n')
+            sys.stderr.write('Zip file does not include %s .\n' %os.path.basename(self.source))
             return False
 
         try:
             self.fn = compile(code, filename = filename, mode = 'exec')
             if 'main' in message:
                 self.mainFn = message['main']
+
+            # if the directory 'virtualenv' is extracted out of a zip file
+            path_to_virtualenv = os.path.dirname(self.source) + '/virtualenv'
+            if os.path.isdir(path_to_virtualenv):
+                # activate the virtualenv
+                activate_this_file = path_to_virtualenv + '/bin/activate_this.py'
+                if os.path.exists(activate_this_file):
+                    execfile(activate_this_file, dict(__file__=activate_this_file))
```

Comment:
execfile does not modify the globals (as __file__) so the executed script may take an incorrect path.
You can pass global variables to execfile so you can modify its __file__ variable